app/src/main/java/com/lagradost/cloudstream3/syncproviders/providers/OpenSubtitlesApi.kt

```diff
@@ -163,17 +164,21 @@ class OpenSubtitlesApi(index: Int) : InAppAuthAPIManager(index), AbstractSubApi
         throwIfCantDoRequest()
         val fixedLang = fixLanguage(query.lang)
 
-        val imdbId = query.imdb ?: 0
         val queryText = query.query.replace(" ", "+")
         val epNum = query.epNumber ?: 0
         val seasonNum = query.seasonNumber ?: 0
         val yearNum = query.year ?: 0
+        val tmdbprov = TmdbProvider()
+        var imdbId = ""
+        if (seasonNum == 0 && epNum == 0) { // applicable only for movies for now
+            imdbId = tmdbprov.getImdb(queryText, yearNum)
```

Comment:
This should be moved up the chain as not only opensubtitles could benefit from IMDb id
See https://github.com/recloudstream/cloudstream/blob/master/app/src/main/java/com/lagradost/cloudstream3/ui/player/GeneratorPlayer.kt#L551

This would of course require adding a get IMDb id method to the viewmodel which would allow for easy caching via an hashmap.

Reply:
You do not want to fetch imdbId twice if you click the load subtitles twice, hence why you want cache and to do that you move the getImdbId method to https://github.com/recloudstream/cloudstream/blob/master/app/src/main/java/com/lagradost/cloudstream3/ui/player/PlayerGeneratorViewModel.kt 

Viewmodels handle all the business logic and allows you to keep a local internal state, e.g a hashmap of previous search results. See https://developer.android.com/topic/architecture/ui-layer/stateholders